proof_of_play.go

```diff
@@ -259,61 +88,70 @@ func (p *proofOfPlay) expire(popReq *PoPRequest) error {
 		}
 	}
 
-	req, err := http.NewRequest("GET", expUrl, nil)
-	if err != nil {
-		return &PoPError{
-			Status:  http.StatusBadRequest,
-			Message: fmt.Sprintf("Connection Error: %s", err.Error()),
-		}
-	}
-
-	return p.makePoPRequest(req, popReq, adId, "expire")
+	return p.confirm(&PoPRequest{
+		Ad:          ad,
+		AdId:        adId,
+		Status:      true,
+		DisplayTime: displayTime,
+		Url:         confirmUrl,
+	})
 }
 
-func (p *proofOfPlay) isLeaseExpired(ad Ad) bool {
-	exp, ok := ad["lease_expiry"].(float64)
-	if !ok {
-		return true
+func (p *proofOfPlay) publishEvent(name string, message string, level string) {
+	if p.eventFn == nil {
+		return
 	}
 
-	expiry := time.Unix(int64(exp), 0)
-	return time.Now().After(expiry)
+	p.eventFn(name, message, "", level)
 }
 
-func (p *proofOfPlay) processRetries() {
-	for req := range p.retryQueue {
-		p.processRetry(req)
+func (p *proofOfPlay) confirm(popReq *PoPRequest) error {
+	data := &ProofOfPlayRequest{DisplayTime: popReq.DisplayTime}
+
+	resp, err := p.popFunc(http.MethodPost, popReq.Url, data)
```

Comment:
this is a bit strange that popFunc returns and error and a resp, we do some stuff based on resp and then return the err. what I'd expect is to check the error and see if we should bother with resp at all. is there a reason why you're doing it this way?

Reply:
I did this way, so that we don't raise an event when there is a connection error, in that case response would be nil, if Its okay to raise event in any failure scenario, I can make the change to check for error and raise the event.